open-craft/edx-platform
# feat: Teaching Assistant role [BB-8580]

## Description

Adds a new `Teaching Assistant` course team role, which is similar to [eSHE Instructor](https://github.com/open-craft/edx-platform/pull/561), with the only difference being that it doesn't have access to viewing and modifying grades.

Provided you set `ENABLE_TEACHING_ASSISTANT_ROLE` feature flag to `True`, as an `Admin` or `Staff`, you should see this on the `Membership` tab of the instructor dashboard:

<details>
  <summary>Screenshot</summary>

![image](https://github.com/open-craft/edx-platform/assets/18251194/46b4fd8d-0798-4bf3-9c9b-93380fa60c88)

</details>

Here is how the dashboard looks for a user with this role:


<details>
  <summary>Screenshot</summary>

  ![image](https://github.com/open-craft/edx-platform/assets/18251194/1979804b-f358-445c-b4aa-0f16c6ce5553)

</details>

You can see that the `Membership` tab is hidden and that the `Student Admin` tab contains only two sections, unlike the regular `Staff`, from which the role in question is derived:

<details>
  <summary>Screenshot</summary>

  ![image](https://github.com/open-craft/edx-platform/assets/18251194/dfad3f70-897b-47e5-8ffa-3dfb30a5e510)
  
</details>

This role should be composable with any other course team roles. For example, `Teaching Assistant` + `Admin` will have the same permissions as `Admin`. `Teaching Assistant` + `Discussion Admin` will be able to access the membership tab, but assign only forum-related roles.

## Testing instructions

1. Switch your devstack to the `0x29a/bb8580/teaching-assistant` branch from the `git@github.com:open-craft/edx-platform.git` repo.
2. Set `ENABLE_TEACHING_ASSISTANT_ROLE` to `True`.
3. Create two users: `course_staff@example.com` and `ta@example.com`. Give the former `Staff` course team role using the Instructor Dashboard.
4. Since we've changed the `instructor.enroll`, `instructor.give_student_extension`, and `instructor.override_grades` permissions, verify that `course_staff@example.com` still can enroll students (`Membership` tab) and that all features on the `Student Admin` tab work for this user as before.
5. Now, give `ta@example.com` the `Teaching Assistant` role.
6. As `ta@example.com`, verify that you have access to the Instructor Dashboard, but not to the `Membership` tab.
7. Also, check that `ta@example.com` sees only two sections on the `Student Admin` tab of the dashboard.
8. Now, give `ta@example.com` the `Staff` role and verify that it has all `Staff` permissions (can enroll students, see the `Membership` tab and all `Student Admin` content).
9. Remove the `Staff` role from `ta@example.com`, give `Discussion Admin` and verify that it can assign forum moderator roles.
10. Remove the `Forum Discussion` role from `ta@example.com`, give `Data Researcher` and verify that it can access the `Data Download` tab.

Additionally, this script can be used to test that the `instructor.enroll`, `instructor.give_student_extension`, and `instructor.override_grades` permissions work as expected:
```python
from django.contrib.auth import get_user_model
from openedx.core.lib.courses import get_course_by_id
from opaque_keys.edx.keys import CourseKey
from lms.djangoapps.courseware.rules import HasAccessRule, HasRolesRule

course_key = CourseKey.from_string("course-v1:Test+Test+Test")
course = get_course_by_id(course_key)
user = get_user_model().objects.get(email="ta@example.com")

assert not user.has_perm('instructor.enroll', course_key)
assert not user.has_perm('instructor.give_student_extension', course_key)
assert not user.has_perm('instructor.override_grades', course_key)
```
